lib/authentication.js

```diff
@@ -0,0 +1,255 @@
+'use strict';
+
+const unicodeCategories = require('./utils/unicode-categories');
+
+const unicodeCategoryIncludes = (categoryName, str, position) => {
+  const charCode = str.charCodeAt(position);
+  const category = unicodeCategories[categoryName];
+  let left = 0;
+  let right = category.length - 1;
+  while (left <= right) {
+    const mid = left + Math.floor((right - left) / 2);
+    if (typeof category[mid] === 'number') {
+      if (category[mid] < charCode) left = mid + 1;
+      else if (category[mid] > charCode) right = mid - 1;
+      else return true;
+    } else if (Array.isArray(category[mid])) {
+      if (category[mid][1] < charCode) left = mid + 1;
+      else if (category[mid][0] > charCode) right = mid - 1;
+      else return true;
+    }
+  }
+  return false;
+};
+
+const passwordTests = {
+  MIN_LENGTH: [
+    (password, options) => password.length >= options.minLength,
+    options => `The password must be longer than ${options.minLength} chars`,
+    { minLength: 10  }
+  ],
+  MAX_LENGTH: [
+    (password, options) => password.length <= options.maxLength,
+    options => `The password must be less than ${options.maxLength} chars`,
+    { maxLength: 128 }
+  ],
+  IS_PHRASE: [
+    (password, options) => password.length >= options.minLength,
+    options => `The passphrase must be longer than ${options.minLength} chars`,
+    { minLength: 20 }
+  ],
+  REPEAT_CHARS: [
+    (password, options) => {
+      const regexp = new RegExp(`(.)\\1{${options.number},}`);
+      return !regexp.test(password);
+    },
+    options => 'The password can not contain a sequence of more than ' +
+      `${options.number} repeat chars`,
+    { number: 2 }
+  ],
+  LOWERCASE_CHARS: [
+    (password, option) => {
+      let lowercaseChars = 0;
+      for (let position = 0; position < password.length; position++) {
+        if (unicodeCategoryIncludes('Ll', password, position)) {
+          if (++lowercaseChars === option.number) return true;
+        }
+      }
+    },
+    options => `The password must contain more than ${options.number} ` +
+      'lowercase chars',
+    { number: 1 }
+  ],
+  UPPERCASE_CHARS: [
+    (password, options) => {
+      let uppercaseChars = 0;
+      for (let position = 0; position < password.length; position++) {
+        if (unicodeCategoryIncludes('Lu', password, position)) {
+          if (++uppercaseChars === options.number) return true;
+        }
+      }
+    },
+    options => `The password must contain more than ${options.number} ` +
+      'uppercase chars',
+    { number: 1 }
+  ],
+  NUMBERS: [
+    (password, options) => {
+      const NUMBERS = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10'];
```

Comment:
I don't think '10' is correct here (at least your check below won't ever check for '10'), is it?